File: Java/week11/collatz/CollatzMain.java
/*
 Filename: CollatzMain.java
 Author: Mikhail Zubko
 Created: 2025/03/27
 Purpose: 
*/

import java.util.ArrayList;
import java.util.Scanner;

public class CollatzMain {
  // Get correct input
  private static int getIntInput(Scanner scanner) {
    while (true) {
      System.out.print("Enter a whole number: ");
      String input = scanner.nextLine().trim().replace(",", "");
      try {
        return Integer.parseInt(input);
      } catch (NumberFormatException e) {
        System.out.println("Wrong input.");
      }
    }
  }

  // Find the number below the given input that produces the longest Collatz chain
  private static void challenge(Collatz collatz, int num) {
    long start = System.nanoTime();

    int maxL = 0;
    int maxN = 0;

    for (int i = 1; i < num; i += 2) {
      int len = collatz.getLen(i);
      if (len > maxL) {
        maxL = len;
        maxN = i;
      }
    }

    System.out.printf("The number %d produces the longest chain of %d steps.%n", maxN, maxL);
    System.out.printf("The function executed in: %.3f seconds%n", (System.nanoTime() - start) / 1e9);
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    Collatz collatz = new Collatz();
    int num;

    System.out.println("The Collatz chain for the entered number.");
    while (true) {
      num = getIntInput(scanner);
      ArrayList<Integer> chain = collatz.getLst(num);

      for (int n : chain) {
        System.out.print(n + " ");
      }
      System.out.println("\nSteps: " + chain.size());
      System.out.print("\nWould you like to try another number? y/n: ");
      String inputstr = scanner.nextLine();
      char ans = inputstr.charAt(0);
      if (ans != 'y') {
        break;
      }
    }

    System.out.println("\nChallenge!\nFind the number below the entered producing the longest Collatz chain.");
    num = getIntInput(scanner);
    challenge(collatz, num);

    scanner.close();
  }
}

File: Cpp/week13/collatz/collatz_main.cpp
#include "collatz.h"
#include <iostream>
#include <limits>
#include <vector>

int get_input() {
  int num{0};
  std::cout << "Enter a whole number: ";
  std::cin >> num;
  // handle wrong input
  if (std::cin.fail()) {
    std::cin.clear(); // clear the error flag
    // discard invalid input
    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    std::cout << " Wrong input." << std::endl;
  }
  return num;
}

int main() {
  std::cout << "\n -- The Collatz chain for the entered number --\n"
            << std::endl;
  Collatz collatz;
  int num;
  int steps;
  char ans;
  std::vector<int> chain;
  while (true) {
    num = get_input();
    chain = collatz.get_lst(num);
    steps = chain.size();
    // print the chain
    for (int element : chain) {
      std::cout << element << " ";
    }
    std::cout << std::endl;
    std::cout << "Steps: " << steps << std::endl;
    std::cout << "Do you want to get another chain? y/n: ";
    std::cin >> ans;
    if (ans != 'y') {
      break;
    }
  }
  std::cout << "Let's try challenge!" << std::endl;
  std::cout << "Find the number below the entered producing the longest "
               "Collatz chain."
            << std::endl;
  num = get_input();
  int maxL{0};
  int maxN{0};
  for (int i = 1; i < num; i += 2) {
    int len = collatz.get_len(i);
    if (len > maxL) {
      maxL = len;
      maxN = i;
    }
  }
  std::cout << "The number " << maxN << " produces the longest chain of "
            << maxL << " steps." << std::endl;
}
